---
title: "Python Packages: Pandas and BioPython"
code-copy: true
jupyter: python3
---

## Pandas 
Python has lots of packages that allows you to manipulate data in different ways. 
For example, Pandas allows you to manipulate dataset with a similar dataframe structure that the language R has. 

Pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. 

Pandas is well suited for many different kinds of data:

- Tabular data with heterogeneously-typed columns, as in an SQL table or Excel spreadsheet

- Ordered and unordered (not necessarily fixed-frequency) time series data.

- Arbitrary matrix data (homogeneously typed or heterogeneous) with row and column labels

- Any other form of observational / statistical data sets. The data need not be labeled at all to be placed into a pandas data structure.


For today we are going to learn basics commands that will allow us to upload data from a CSV, perform commands that will allow us to get general data from the dataset and filter by column. 

### Importing the Pandas package
To work woth any package in python you have to download it through your package manager (in the case of this workshop would be Conda), and then import it. For this workshop, the conda enviroment already had installed Pandas. 

The following code will allow you to import the package Pandas to your 
```{python}
import pandas as pd
```

### How to load a CSV into Pandas

```{python}
df = pd.read_csv('../python_lesson/combined_tidy_vcf.csv')
```

Note: The directory path were the dataset is located may be in a diferent place than this example. 

### Getting information about the dataset

```{python}
df.info()
```

### Getting the first rows of the dataset

```{python}
df.head()
```

### Getting the last rows of the dataset

```{python}
df.tail()
```

### Getting the name of the columns

```{python}
df.columns
```

### Identifiying Null values

```{python}
df.isnull()
```

### Identifiying Null values and counting them

```{python}
df.isnull().sum()
```

### Filtering a column

In this example we are going to filter the dataset to alter allele sequences of at least length 10. To do this we need to apply a transformation to the `ALT` column using `apply`.

```{python}
df['ALT'].apply(len)
```

This vector of lengths can then be compared to a minimum length to negerate a boolean vector.

```{python}
min_len = 10
df['ALT'].apply(len) >= min_len
```

Using this boolean vector we can select only the rows that met the condition by using the [loc](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html) method. 

```{python}
df.loc[df['ALT'].apply(len) >= min_len]
```

## BioPython
BioPython is another packages that Python includes. 
It features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.

### To import BioPython Package

For this exersise we are going to import two BioPython (or simply `Bio`) sub-modules `Seq` and `pairwise2`.

```{python}
from Bio import Seq, pairwise2
```

The [Seq](https://biopython.org/docs/1.75/api/Bio.Seq.html) module has a sequence object that stores sequence data in memory more efficiently than strings. 

The [pairwise2](https://biopython.org/docs/1.75/api/Bio.pairwise2.html) module has functions that can be used to compare sequences.

### Gathering sequences from the dataset we loaded using pandas

Select one of the rows found by the filter and extract the reference and the alter sequences.

```{python}
row = 8
ref = Seq.Seq(df.loc[row]['REF'])
alt = Seq.Seq(df.loc[row]['ALT'])

print(ref)
print(alt)
```

### Auto Generate Alignments between the reference and the altered sequence

We can generate alignments between the reference and alter sequences using the `pairwise2.align.globalxx` function.

```{python}
alignments = pairwise2.align.globalxx(ref,alt)
print(len(alignments))
```

To visualize this alignments, pairwise2 provides the [format_alignment](https://biopython.org/docs/1.75/api/Bio.pairwise2.html#Bio.pairwise2.format_alignment) function that needs to receive multiple parameters. Thankfully

```{python}
for alignment in alignments:
    print(pairwise2.format_alignment(*alignment))
```





