{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: Redirection\n",
        "format: html\n",
        "engine: jupyter\n",
        "---"
      ],
      "id": "5918bb75"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "::::::::::::::::::::::::::::::::::::::: objectives\n",
        "\n",
        "- Employ the `grep` command to search for information within files.\n",
        "- Print the results of a command to a file.\n",
        "- Construct command pipelines with two or more stages.\n",
        "- Use `for` loops to run the same command for several input files.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::: questions\n",
        "\n",
        "- How can I search within files?\n",
        "- How can I combine existing commands to do new things?\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "## Searching files\n",
        "\n",
        "We discussed in a previous episode how to search within a file using `less`. We can also\n",
        "search within files without even opening them, using `grep`. `grep` is a command-line\n",
        "utility for searching plain-text files for lines matching a specific set of\n",
        "characters (sometimes called a string) or a particular pattern\n",
        "(which can be specified using something called regular expressions). We're not going to work with\n",
        "regular expressions in this lesson, and are instead going to specify the strings\n",
        "we are searching for.\n",
        "Let's give it a try!\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## Nucleotide abbreviations\n",
        "\n",
        "The four nucleotides that appear in DNA are abbreviated `A`, `C`, `T` and `G`.\n",
        "Unknown nucleotides are represented with the letter `N`. An `N` appearing\n",
        "in a sequencing file represents a position where the sequencing machine was not able to\n",
        "confidently determine the nucleotide in that position. You can think of an `N` as being aNy\n",
        "nucleotide at that position in the DNA sequence.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "We'll search for strings inside of our fastq files. Let's first make sure we are in the correct\n",
        "directory:\n",
        "\n",
        "\n",
        "```{bash}\n",
        "cd ~/shell_data/untrimmed_fastq\n",
        "```\n",
        "\n",
        "\n",
        "Suppose we want to see how many reads in our file have really bad segments containing 10 consecutive unknown nucleotides (Ns).\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## Determining quality\n",
        "\n",
        "In this lesson, we're going to be manually searching for strings of `N`s within our sequence\n",
        "results to illustrate some principles of file searching. It can be really useful to do this\n",
        "type of searching to get a feel for the quality of your sequencing results, however, in your\n",
        "research you will most likely use a bioinformatics tool that has a built-in program for\n",
        "filtering out low-quality reads. You'll learn how to use one such tool in\n",
        "[a later lesson](https://datacarpentry.org/wrangling-genomics/02-quality-control).\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "Let's search for the string NNNNNNNNNN in the SRR098026 file:\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep NNNNNNNNNN SRR098026.fastq\n",
        "```\n",
        "\n",
        "\n",
        "This command returns a lot of output to the terminal. Every single line in the SRR098026\n",
        "file that contains at least 10 consecutive Ns is printed to the terminal, regardless of how long or short the file is.\n",
        "We may be interested not only in the actual sequence which contains this string, but\n",
        "in the name (or identifier) of that sequence. We discussed in a previous lesson\n",
        "that the identifier line immediately precedes the nucleotide sequence for each read\n",
        "in a FASTQ file. We may also want to inspect the quality scores associated with\n",
        "each of these reads. To get all of this information, we will return the line\n",
        "immediately before each match and the two lines immediately after each match.\n",
        "\n",
        "We can use the `-B` argument for grep to return a specific number of lines before\n",
        "each match. The `-A` argument returns a specific number of lines after each matching line. Here we want the line *before* and the two lines *after* each\n",
        "matching line, so we add `-B1 -A2` to our grep command:\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq\n",
        "```\n",
        "\n",
        "\n",
        "One of the sets of lines returned by this command is:\n",
        "\n",
        "```output\n",
        "@SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "CNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
        "+SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
        "```\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::  challenge\n",
        "\n",
        "## Exercise\n",
        "\n",
        "1. Search for the sequence `GNATNACCACTTCC` in the `SRR098026.fastq` file.\n",
        "  Have your search return all matching lines and the name (or identifier) for each sequence\n",
        "  that contains a match.\n",
        "\n",
        "2. Search for the sequence `AAGTT` in both FASTQ files.\n",
        "  Have your search return all matching lines and the name (or identifier) for each sequence\n",
        "  that contains a match.\n",
        "\n",
        ":::::::::::::::  solution\n",
        "\n",
        "## Solution\n",
        "\n",
        "1. `grep -B1 GNATNACCACTTCC SRR098026.fastq`\n",
        "\n",
        "```\n",
        "@SRR098026.245 HWUSI-EAS1599_1:2:1:2:801 length=35\n",
        "GNATNACCACTTCCAGTGCTGANNNNNNNGGGATG\n",
        "```\n",
        "\n",
        "2. `grep -B1 AAGTT *.fastq`\n",
        "\n",
        "```\n",
        "SRR097977.fastq-@SRR097977.11 209DTAAXX_Lenski2_1_7:8:3:247:351 length=36\n",
        "SRR097977.fastq:GATTGCTTTAATGAAAAAGTCATATAAGTTGCCATG\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.67 209DTAAXX_Lenski2_1_7:8:3:544:566 length=36\n",
        "SRR097977.fastq:TTGTCCACGCTTTTCTATGTAAAGTTTATTTGCTTT\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.68 209DTAAXX_Lenski2_1_7:8:3:724:110 length=36\n",
        "SRR097977.fastq:TGAAGCCTGCTTTTTTATACTAAGTTTGCATTATAA\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.80 209DTAAXX_Lenski2_1_7:8:3:258:281 length=36\n",
        "SRR097977.fastq:GTGGCGCTGCTGCATAAGTTGGGTTATCAGGTCGTT\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.92 209DTAAXX_Lenski2_1_7:8:3:353:318 length=36\n",
        "SRR097977.fastq:GGCAAAATGGTCCTCCAGCCAGGCCAGAAGCAAGTT\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.139 209DTAAXX_Lenski2_1_7:8:3:703:655 length=36\n",
        "SRR097977.fastq:TTTATTTGTAAAGTTTTGTTGAAATAAGGGTTGTAA\n",
        "--\n",
        "SRR097977.fastq-@SRR097977.238 209DTAAXX_Lenski2_1_7:8:3:592:919 length=36\n",
        "SRR097977.fastq:TTCTTACCATCCTGAAGTTTTTTCATCTTCCCTGAT\n",
        "--\n",
        "SRR098026.fastq-@SRR098026.158 HWUSI-EAS1599_1:2:1:1:1505 length=35\n",
        "SRR098026.fastq:GNNNNNNNNCAAAGTTGATCNNNNNNNNNTGTGCG\n",
        "```\n",
        "\n",
        ":::::::::::::::::::::::::\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "## Redirecting output\n",
        "\n",
        "`grep` allowed us to identify sequences in our FASTQ files that match a particular pattern.\n",
        "All of these sequences were printed to our terminal screen, but in order to work with these\n",
        "sequences and perform other operations on them, we will need to capture that output in some\n",
        "way.\n",
        "\n",
        "We can do this with something called \"redirection\". The idea is that\n",
        "we are taking what would ordinarily be printed to the terminal screen and redirecting it to another location.\n",
        "In our case, we want to print this information to a file so that we can look at it later and\n",
        "use other commands to analyze this data.\n",
        "\n",
        "The command for redirecting output to a file is `>`.\n",
        "\n",
        "Let's try out this command and copy all the records (including all four lines of each record)\n",
        "in our FASTQ files that contain\n",
        "'NNNNNNNNNN' to another file called `bad_reads.txt`.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq > bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## File extensions\n",
        "\n",
        "You might be confused about why we're naming our output file with a `.txt` extension. After all,\n",
        "it will be holding FASTQ formatted data that we're extracting from our FASTQ files. Won't it\n",
        "also be a FASTQ file? The answer is, yes - it will be a FASTQ file and it would make sense to\n",
        "name it with a `.fastq` extension. However, using a `.fastq` extension will lead us to problems\n",
        "when we move to using wildcards later in this episode. We'll point out where this becomes\n",
        "important. For now, it's good that you're thinking about file extensions!\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "The prompt should sit there a little bit, and then it should look like nothing\n",
        "happened. But type `ls`. You should see a new file called `bad_reads.txt`.\n",
        "\n",
        "We can check the number of lines in our new file using a command called `wc`.\n",
        "`wc` stands for **word count**. This command counts the number of words, lines, and characters\n",
        "in a file. The FASTQ file may change over time, so given the potential for updates,\n",
        "make sure your file matches your instructor's output.\n",
        "\n",
        "As of Sept. 2020, wc gives the following output:\n",
        "\n",
        "\n",
        "```{bash}\n",
        "wc bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "  802    1338   24012 bad_reads.txt\n",
        "```\n",
        "\n",
        "This will tell us the number of lines, words and characters in the file. If we\n",
        "want only the number of lines, we can use the `-l` flag for `lines`.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "802 bad_reads.txt\n",
        "```\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::  challenge\n",
        "\n",
        "## Exercise\n",
        "\n",
        "How many sequences are there in `SRR098026.fastq`? Remember that every sequence is formed by four lines.\n",
        "\n",
        ":::::::::::::::  solution\n",
        "\n",
        "## Solution\n",
        "\n",
        "\n",
        "```{bash}\n",
        "wc -l SRR098026.fastq\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "996\n",
        "```\n",
        "\n",
        "Now you can divide this number by four to get the number of sequences in your fastq file.\n",
        "\n",
        ":::::::::::::::::::::::::\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::  challenge\n",
        "\n",
        "## Exercise\n",
        "\n",
        "How many sequences in `SRR098026.fastq` contain at least 3 consecutive Ns?\n",
        "\n",
        ":::::::::::::::  solution\n",
        "\n",
        "## Solution\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep NNN SRR098026.fastq > bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "249\n",
        "```\n",
        "\n",
        ":::::::::::::::::::::::::\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "We might want to search multiple FASTQ files for sequences that match our search pattern.\n",
        "However, we need to be careful, because each time we use the `>` command to redirect output\n",
        "to a file, the new output will replace the output that was already present in the file.\n",
        "This is called \"overwriting\" and, just like you don't want to overwrite your video recording\n",
        "of your kid's first birthday party, you also want to avoid overwriting your data files.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq > bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "802 bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR097977.fastq > bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "0 bad_reads.txt\n",
        "```\n",
        "\n",
        "Here, the output of our second  call to `wc` shows that we no longer have any lines in our `bad_reads.txt` file. This is\n",
        "because the second file we searched (`SRR097977.fastq`) does not contain any lines that match our\n",
        "search sequence. So our file was overwritten and is now empty.\n",
        "\n",
        "We can avoid overwriting our files by using the command `>>`. `>>` is known as the \"append redirect\" and will\n",
        "append new output to the end of a file, rather than overwriting it.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq > bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "802 bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR097977.fastq >> bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "802 bad_reads.txt\n",
        "```\n",
        "\n",
        "The output of our second call to `wc` shows that we have not overwritten our original data.\n",
        "\n",
        "We can also do this with a single line of code by using a wildcard:\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN *.fastq > bad_reads.txt\n",
        "wc -l bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "802 bad_reads.txt\n",
        "```\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## File extensions - part 2\n",
        "\n",
        "This is where we would have trouble if we were naming our output file with a `.fastq` extension.\n",
        "If we already had a file called `bad_reads.fastq` (from our previous `grep` practice)\n",
        "and then ran the command above using a `.fastq` extension instead of a `.txt` extension, `grep`\n",
        "would give us a warning.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN *.fastq > bad_reads.fastq\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "grep: input file â€˜bad_reads.fastq' is also the output\n",
        "```\n",
        "\n",
        "`grep` is letting you know that the output file `bad_reads.fastq` is also included in your\n",
        "`grep` call because it matches the `*.fastq` pattern. Be careful with this as it can lead to\n",
        "some unintended results.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "Since we might have multiple different criteria we want to search for,\n",
        "creating a new output file each time has the potential to clutter up our workspace. We also\n",
        "thus far haven't been interested in the actual contents of those files, only in the number of\n",
        "reads that we've found. We created the files to store the reads and then counted the lines in\n",
        "the file to see how many reads matched our criteria. There's a way to do this, however, that\n",
        "doesn't require us to create these intermediate files - the pipe command (`|`).\n",
        "\n",
        "This is probably not a key on\n",
        "your keyboard you use very much, so let's all take a minute to find that key.\n",
        "In the UK and US keyboard layouts, and several others,\n",
        "the `|` character can be found using the key combination <kbd>Shift</kbd>\\+<kbd>\\</kbd>.\n",
        "This may be different for other language-specific layouts.\n",
        "\n",
        "What `|` does is take the output that is scrolling by on the terminal and uses that output as input to another command.\n",
        "When our output was scrolling by, we might have wished we could slow it down and\n",
        "look at it, like we can with `less`. Well it turns out that we can! We can redirect our output\n",
        "from our `grep` call through the `less` command.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq | less\n",
        "```\n",
        "\n",
        "\n",
        "We can now see the output from our `grep` call within the `less` interface. We can use the up and down arrows\n",
        "to scroll through the output and use `q` to exit `less`.\n",
        "\n",
        "If we don't want to create a file before counting lines of output from our `grep` search, we could directly pipe\n",
        "the output of the grep search to the command `wc -l`. This can be helpful for investigating your output if you are not sure\n",
        "you would like to save it to a file.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq | wc -l \n",
        "```\n",
        "\n",
        "\n",
        "Because we asked `grep` for all four lines of each FASTQ record, we need to divide the output by\n",
        "four to get the number of sequences that match our search pattern. Since 802 / 4 = 200.5 and we\n",
        "are expecting an integer number of records, there is something added or missing in `bad_reads.txt`.\n",
        "If we explore `bad_reads.txt` using `less`, we might be able to notice what is causing the uneven\n",
        "number of lines. Luckily, this issue happens by the end of the file so we can also spot it with `tail`.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq > bad_reads.txt\n",
        "tail bad_reads.txt\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "@SRR098026.133 HWUSI-EAS1599_1:2:1:0:1978 length=35\n",
        "ANNNNNNNNNTTCAGCGACTNNNNNNNNNNGTNGN\n",
        "+SRR098026.133 HWUSI-EAS1599_1:2:1:0:1978 length=35\n",
        "#!!!!!!!!!##########!!!!!!!!!!##!#!\n",
        "--\n",
        "--\n",
        "@SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "CNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
        "+SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
        "```\n",
        "\n",
        "The fifth and six lines in the output display \"--\" which is the default action for `grep` to separate groups of\n",
        "lines matching the pattern, and indicate groups of lines which did not match the pattern so are not displayed.\n",
        "To fix this issue, we can redirect the output of grep to a second instance of `grep` as follows.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "grep -B1 -A2 NNNNNNNNNN SRR098026.fastq | grep -v '^--' > bad_reads.fastq\n",
        "tail bad_reads.fastq\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "+SRR098026.132 HWUSI-EAS1599_1:2:1:0:320 length=35\n",
        "#!!!!!!!!!##########!!!!!!!!!!##!#!\n",
        "@SRR098026.133 HWUSI-EAS1599_1:2:1:0:1978 length=35\n",
        "ANNNNNNNNNTTCAGCGACTNNNNNNNNNNGTNGN\n",
        "+SRR098026.133 HWUSI-EAS1599_1:2:1:0:1978 length=35\n",
        "#!!!!!!!!!##########!!!!!!!!!!##!#!\n",
        "@SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "CNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
        "+SRR098026.177 HWUSI-EAS1599_1:2:1:1:2025 length=35\n",
        "#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
        "```\n",
        "\n",
        "The `-v` option in the second `grep` search stands for `--invert-match` meaning `grep` will now only display the\n",
        "lines which do not match the searched pattern, in this case `'^--'`. The caret (`^`) is an **anchoring**\n",
        "character matching the beginning of the line, and the pattern has to be enclose by single quotes so `grep` does\n",
        "not interpret the pattern as an extended option (starting with --).\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## Custom `grep` control\n",
        "\n",
        "Use `man grep` to read more about other options to customize the output of `grep` including extended options,\n",
        "anchoring characters, and much more.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "Redirecting output is often not intuitive, and can take some time to get used to. Once you're\n",
        "comfortable with redirection, however, you'll be able to combine any number of commands to\n",
        "do all sorts of exciting things with your data!\n",
        "\n",
        "None of the command line programs we've been learning\n",
        "do anything all that impressive on their own, but when you start chaining\n",
        "them together, you can do some really powerful things very\n",
        "efficiently.\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::::  callout\n",
        "\n",
        "## File manipulation and more practices with pipes\n",
        "\n",
        "To practice a bit more with the tools we've added to our tool kit so far and learn a few extra ones you can follow [this extra lesson](Extra_lesson.md) which uses the SRA metadata file.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "## Writing for loops\n",
        "\n",
        "Loops are key to productivity improvements through automation as they allow us to execute commands repeatedly.\n",
        "Similar to wildcards and tab completion, using loops also reduces the amount of typing (and typing mistakes).\n",
        "Loops are helpful when performing operations on groups of sequencing files, such as unzipping or trimming multiple\n",
        "files. We will use loops for these purposes in subsequent analyses, but will cover the basics of them for now.\n",
        "\n",
        "When the shell sees the keyword `for`, it knows to repeat a command (or group of commands) once for each item in a list.\n",
        "Each time the loop runs (called an iteration), an item in the list is assigned in sequence to the **variable**, and\n",
        "the commands inside the loop are executed, before moving on to the next item in the list. Inside the loop, we call for\n",
        "the variable's value by putting `$` in front of it. The `$` tells the shell interpreter to treat the **variable**\n",
        "as a variable name and substitute its value in its place, rather than treat it as text or an external command. In shell programming, this is usually called \"expanding\" the variable.\n",
        "\n",
        "Sometimes, we want to expand a variable without any whitespace to its right.\n",
        "Suppose we have a variable named `foo` that contains the text `abc`, and would\n",
        "like to expand `foo` to create the text `abcEFG`.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "foo=abc\n",
        "echo foo is $foo\n",
        "foo is abc\n",
        "echo foo is $fooEFG      # doesn't work\n",
        "foo is\n",
        "```\n",
        "\n",
        "\n",
        "The interpreter is trying to expand a variable named `fooEFG`, which (probably)\n",
        "doesn't exist. We can avoid this problem by enclosing the variable name in\n",
        "braces (`{` and `}`, also called \"curly brackets\"). `bash` treats the `#`\n",
        "character as a comment character. Any text on a line after a `#` is ignored by\n",
        "bash when evaluating the text as code.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "foo=abc\n",
        "echo foo is $foo\n",
        "echo foo is ${foo}EFG      # now it ```\n",
        "\n",
        "Let's write a for loop to show us the first two lines of the fastq files we downloaded earlier. You will notice the shell prompt changes from `$` to `>` and back again as we were typing in our loop. The second prompt, `>`, is different to remind us that we haven't finished typing a complete command yet. A semicolon, `;`, can be used to separate two commands written on a single line.\n",
        "\n",
        "```{bash}\n",
        "cd ../untrimmed_fastq/\n",
        "```\n",
        "\n",
        "```{bash}\n",
        "for filename in *.fastq\n",
        "> do\n",
        "> head -n 2 ${filename}\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        "The for loop begins with the formula `for <variable> in <group to iterate over>`. In this case, the word `filename` is designated\n",
        "as the variable to be used over each iteration. In our case `SRR097977.fastq` and `SRR098026.fastq` will be substituted for `filename`\n",
        "because they fit the pattern of ending with .fastq in the directory we've specified. The next line of the for loop is `do`. The next line is\n",
        "the code that we want to execute. We are telling the loop to print the first two lines of each variable we iterate over. Finally, the\n",
        "word `done` ends the loop.\n",
        "\n",
        "After executing the loop, you should see the first two lines of both fastq files printed to the terminal. Let's create a loop that\n",
        "will save this information to a file.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "for filename in *.fastq\n",
        "> do\n",
        "> head -n 2 ${filename} >> seq_info.txt\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        "When writing a loop, you will not be able to return to previous lines once you have pressed Enter. Remember that we can cancel the current command using\n",
        "\n",
        "- <kbd>Ctrl</kbd>\\+<kbd>C</kbd>\n",
        "\n",
        "If you notice a mistake that is going to prevent your loop for executing correctly.\n",
        "\n",
        "Note that we are using `>>` to append the text to our `seq_info.txt` file. If we used `>`, the `seq_info.txt` file would be rewritten\n",
        "every time the loop iterates, so it would only have text from the last variable used. Instead, `>>` adds to the end of the file.\n",
        "\n",
        "## Using Basename in for loops\n",
        "\n",
        "Basename is a function in UNIX that is helpful for removing a uniform part of a name from a list of files. In this case, we will use basename to remove the `.fastq` extension from the files that we've been working with.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "basename SRR097977.fastq .fastq\n",
        "```\n",
        "\n",
        "\n",
        "We see that this returns just the SRR accession, and no longer has the .fastq file extension on it.\n",
        "\n",
        "```output\n",
        "SRR097977\n",
        "```\n",
        "\n",
        "If we try the same thing but use `.fasta` as the file extension instead, nothing happens. This is because basename only works when it exactly matches a string in the file.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "basename SRR097977.fastq .fasta\n",
        "```\n",
        "\n",
        "\n",
        "```output\n",
        "SRR097977.fastq\n",
        "```\n",
        "\n",
        "Basename is really powerful when used in a for loop. It allows to access just the file prefix, which you can use to name things. Let's try this.\n",
        "\n",
        "Inside our for loop, we create a new name variable. We call the basename function inside the parenthesis, then give our variable name from the for loop, in this case `${filename}`, and finally state that `.fastq` should be removed from the file name. It's important to note that we're not changing the actual files, we're creating a new variable called name. The line > echo $name will print to the terminal the variable name each time the for loop runs. Because we are iterating over two files, we expect to see two lines of output.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "for filename in *.fastq\n",
        "> do\n",
        "> name=$(basename ${filename} .fastq)\n",
        "> echo ${name}\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::  challenge\n",
        "\n",
        "## Exercise\n",
        "\n",
        "Print the file prefix of all of the `.txt` files in our current directory.\n",
        "\n",
        ":::::::::::::::  solution\n",
        "\n",
        "## Solution\n",
        "\n",
        "\n",
        "```{bash}\n",
        "for filename in *.txt\n",
        "> do\n",
        "> name=$(basename ${filename} .txt)\n",
        "> echo ${name}\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        ":::::::::::::::::::::::::\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        "One way this is really useful is to move files. Let's rename all of our .txt files using `mv` so that they have the years on them, which will document when we created them.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "for filename in *.txt\n",
        "> do\n",
        "> name=$(basename ${filename} .txt)\n",
        "> mv ${filename}  ${name}_2019.txt\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::  challenge\n",
        "\n",
        "## Exercise\n",
        "\n",
        "Remove `_2019` from all of the `.txt` files.\n",
        "\n",
        ":::::::::::::::  solution\n",
        "\n",
        "## Solution\n",
        "\n",
        "\n",
        "```{bash}\n",
        "for filename in *_2019.txt\n",
        "> do\n",
        "> name=$(basename ${filename} _2019.txt)\n",
        "> mv ${filename} ${name}.txt\n",
        "> done\n",
        "```\n",
        "\n",
        "\n",
        ":::::::::::::::::::::::::\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n",
        "\n",
        ":::::::::::::::::::::::::::::::::::::::: keypoints\n",
        "\n",
        "- `grep` is a powerful search tool with many options for customization.\n",
        "- `>`, `>>`, and `|` are different ways of redirecting output.\n",
        "- `command > file` redirects a command's output to a file.\n",
        "- `command >> file` redirects a command's output to a file without overwriting the existing contents of the file.\n",
        "- `command_1 | command_2` redirects the output of the first command as input to the second command.\n",
        "- `for` loops are used for iteration.\n",
        "- `basename` gets rid of repetitive parts of names.\n",
        "\n",
        "::::::::::::::::::::::::::::::::::::::::::::::::::\n"
      ],
      "id": "bb0da683"
    }
  ],
  "metadata": {
    "kernelspec": {
      "name": "python3",
      "language": "python",
      "display_name": "Python 3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}