<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.475">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>CDABS Intro to Programming - Installing Shell Software</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">CDABS Intro to Programming</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-shell" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Shell</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-shell">    
        <li>
    <a class="dropdown-item" href="../shell_lesson/06-installing-shell-software.html">
 <span class="dropdown-text">Installing Shell Software</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell_lesson/01-introduction.html">
 <span class="dropdown-text">Introducing the Shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell_lesson/02-the-filesystem.html">
 <span class="dropdown-text">Navigating Files and Directories</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell_lesson/04-redirection.html">
 <span class="dropdown-text">Redirection</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell_lesson/03-working-with-files.html">
 <span class="dropdown-text">Working with Files and Directories</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell_lesson/05-writing-scripts.html">
 <span class="dropdown-text">Writing Scripts and Working with Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">    
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">    
        <li>
    <a class="dropdown-item" href="../python_lesson/13-files.html">
 <span class="dropdown-text">Analyzing Data from Multiple Files</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/16-cmdline.html">
 <span class="dropdown-text">Command-Line Programs</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/09-func.html">
 <span class="dropdown-text">Creating Functions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/10-errors.html">
 <span class="dropdown-text">Errors and Exceptions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/14-filtering.html">
 <span class="dropdown-text">Filtering Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/17-conda.html">
 <span class="dropdown-text">Installing Bioinformatics Packages</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/01-intro.html">
 <span class="dropdown-text">Introduction to Python</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/06-conditionals.html">
 <span class="dropdown-text">Making Choices</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/03-operators.html">
 <span class="dropdown-text">Operators</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/12-plotting.html">
 <span class="dropdown-text">Producing Plots</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/11-loaddata.html">
 <span class="dropdown-text">Reading in Tabular Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/07-loop.html">
 <span class="dropdown-text">Repeating Actions with Loops</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/04-collections.html">
 <span class="dropdown-text">Storing Multiple Values in Collections</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/15-tidying.html">
 <span class="dropdown-text">Tidying up the Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/02-variables.html">
 <span class="dropdown-text">Types, Objects, Values and Variables</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/08-libraries.html">
 <span class="dropdown-text">Using External Modules</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../python_lesson/05-scripts.html">
 <span class="dropdown-text">Writing Python Scripts</span></a>
  </li>  
    </ul>
  </li>
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Installing Shell Software</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/06-installing-shell-software.html" class="sidebar-item-text sidebar-link active">Installing Shell Software</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/01-introduction.html" class="sidebar-item-text sidebar-link">Introducing the Shell</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/02-the-filesystem.html" class="sidebar-item-text sidebar-link">Navigating Files and Directories</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/04-redirection.html" class="sidebar-item-text sidebar-link">Redirection</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/03-working-with-files.html" class="sidebar-item-text sidebar-link">Working with Files and Directories</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../shell_lesson/05-writing-scripts.html" class="sidebar-item-text sidebar-link">Writing Scripts and Working with Data</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul class="collapse">
  <li><a href="#software" id="toc-software" class="nav-link" data-scroll-target="#software">Software</a></li>
  <li><a href="#quickstart-software-installation-instructions" id="toc-quickstart-software-installation-instructions" class="nav-link" data-scroll-target="#quickstart-software-installation-instructions">QuickStart Software Installation Instructions</a></li>
  <li><a href="#miniconda3" id="toc-miniconda3" class="nav-link" data-scroll-target="#miniconda3">Miniconda3</a></li>
  <li><a href="#fastqc" id="toc-fastqc" class="nav-link" data-scroll-target="#fastqc">FastQC</a></li>
  <li><a href="#trimmomatic" id="toc-trimmomatic" class="nav-link" data-scroll-target="#trimmomatic">Trimmomatic</a></li>
  <li><a href="#bwa" id="toc-bwa" class="nav-link" data-scroll-target="#bwa">BWA</a></li>
  <li><a href="#samtools" id="toc-samtools" class="nav-link" data-scroll-target="#samtools">SAMtools</a></li>
  <li><a href="#bcftools" id="toc-bcftools" class="nav-link" data-scroll-target="#bcftools">BCFtools</a></li>
  <li><a href="#igv" id="toc-igv" class="nav-link" data-scroll-target="#igv">IGV</a></li>
  <li><a href="#running-fastqc" id="toc-running-fastqc" class="nav-link" data-scroll-target="#running-fastqc">Running FastQC</a></li>
  <li><a href="#viewing-the-fastqc-results" id="toc-viewing-the-fastqc-results" class="nav-link" data-scroll-target="#viewing-the-fastqc-results">Viewing the FastQC results</a></li>
  <li><a href="#assessing-quality-using-fastqc" id="toc-assessing-quality-using-fastqc" class="nav-link" data-scroll-target="#assessing-quality-using-fastqc">Assessing Quality using FastQC</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Installing Shell Software</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="overview" class="level1">
<h1>Overview</h1>
<p>We have gone over many ways that you can use the shell to control your computer using commands entered with a keyboard instead of controlling graphical user interfaces (GUIs) with a mouse/keyboard combination.</p>
<p>One of the main reasons we spend so much time learning all these fundamentals is to be able to apply them in practice. One way which you can apply these is in combination with running software from the shell. So far all the commands you have been working with are part of BASH. There are programs that are designed to be run from the shell specifically. The following table provides some examples of commonly used command line software tools.</p>
<section id="software" class="level3">
<h3 class="anchored" data-anchor-id="software">Software</h3>
<table class="table">
<thead>
<tr class="header">
<th>Software</th>
<th>Version</th>
<th>Manual</th>
<th>Available for</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td>
<td>0.11.7</td>
<td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>Quality control tool for high throughput sequence data.</td>
</tr>
<tr class="even">
<td><a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></td>
<td>0.38</td>
<td><a href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>A flexible read trimming tool for Illumina NGS data.</td>
</tr>
<tr class="odd">
<td><a href="http://bio-bwa.sourceforge.net/">BWA</a></td>
<td>0.7.17</td>
<td><a href="http://bio-bwa.sourceforge.net/bwa.shtml">Link</a></td>
<td>Linux, MacOS</td>
<td>Mapping DNA sequences against reference genome.</td>
</tr>
<tr class="even">
<td><a href="http://samtools.sourceforge.net/">SAMtools</a></td>
<td>1.9</td>
<td><a href="http://www.htslib.org/doc/samtools.html">Link</a></td>
<td>Linux, MacOS</td>
<td>Utilities for manipulating alignments in the SAM format.</td>
</tr>
<tr class="odd">
<td><a href="https://samtools.github.io/bcftools/">BCFtools</a></td>
<td>1.8</td>
<td><a href="https://samtools.github.io/bcftools/bcftools.html">Link</a></td>
<td>Linux, MacOS</td>
<td>Utilities for variant calling and manipulating VCFs and BCFs.</td>
</tr>
<tr class="even">
<td><a href="http://software.broadinstitute.org/software/igv/home">IGV</a></td>
<td><a href="https://software.broadinstitute.org/software/igv/download">Link</a></td>
<td><a href="https://software.broadinstitute.org/software/igv/UserGuide">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>Visualization and interactive exploration of large genomics datasets.</td>
</tr>
</tbody>
</table>
</section>
<section id="quickstart-software-installation-instructions" class="level3">
<h3 class="anchored" data-anchor-id="quickstart-software-installation-instructions">QuickStart Software Installation Instructions</h3>
<p>These are the QuickStart installation instructions. They assume familiarity with the command line and with installation in general. As there are different operating systems and many different versions of operating systems and environments, these may not work on your computer. If an installation doesn’t work for you, please refer to the user guide for the tool, listed in the table above.</p>
<p>We have installed software using <a href="https://docs.conda.io/en/latest/miniconda.html">miniconda</a>. Miniconda is a package manager that simplifies the installation process. Please first install miniconda3 (installation instructions below), and then proceed to the installation of individual tools.</p>
</section>
<section id="miniconda3" class="level3">
<h3 class="anchored">Miniconda3</h3>
<blockquote class="blockquote">
<h2 id="macos" class="anchored" data-anchor-id="miniconda3">MacOS</h2>
<p>To install miniconda3, type:</p>
<pre><code>$ curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
$ bash Miniconda3-latest-MacOSX-x86_64.sh</code></pre>
<p>{: .bash} Then, follow the instructions that you are prompted with on the screen to install Miniconda3. {: .solution}</p>
</blockquote>
</section>
<section id="fastqc" class="level3">
<h3 class="anchored">FastQC</h3>
<blockquote class="blockquote">
<h2 id="macos-1" class="anchored" data-anchor-id="fastqc">MacOS</h2>
<p>To install FastQC, type:</p>
<pre><code>$ conda install -c bioconda fastqc=0.11.7=5</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="fastqc-source-code-installation" class="anchored">FastQC Source Code Installation</h2>
<p>If you prefer to install from source, follow the directions below:</p>
<pre><code>$ cd ~/src
$ curl -O http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.7.zip
$ unzip fastqc_v0.11.7.zip</code></pre>
<p>{: .bash}</p>
<p>Link the fastqc executable to the ~/bin folder that you have already added to the path.</p>
<pre><code>$ ln -sf ~/src/FastQC/fastqc ~/bin/fastqc</code></pre>
<p>{: .bash}</p>
<p>Due to what seems a packaging error the executable flag on the fastqc program is not set. We need to set it ourselves.</p>
<pre><code>$ chmod +x ~/bin/fastqc</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<p><strong>Test your installation by running:</strong></p>
<pre><code>$ fastqc -h</code></pre>
<p>{: .bash}</p>
</section>
<section id="trimmomatic" class="level3">
<h3 class="anchored">Trimmomatic</h3>
<blockquote class="blockquote">
<h2 id="macos-2" class="anchored" data-anchor-id="trimmomatic">MacOS</h2>
<pre><code>conda install -c bioconda trimmomatic=0.38=0</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="trimmomatic-source-code-installation" class="anchored">Trimmomatic Source Code Installation</h2>
<p>If you prefer to install from source, follow the directions below:</p>
<pre><code>$ cd ~/src
$ curl -O http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.38.zip
$ unzip Trimmomatic-0.38.zip</code></pre>
<p>{: .bash}</p>
<p>The program can be invoked via:</p>
<pre><code>$ java -jar ~/src/Trimmomatic-0.38/trimmomatic-0.38.jar</code></pre>
<p>The ~/src/Trimmomatic-0.38/adapters/ directory contains Illumina specific adapter sequences.</p>
<pre><code>$ ls ~/src/Trimmomatic-0.38/adapters/</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<p><strong>Test your installation by running:</strong> (assuming things are installed in ~/src)</p>
<pre><code>$ java -jar ~/src/Trimmomatic-0.38/trimmomatic-0.38.jar</code></pre>
<p>{: .bash}</p>
<blockquote class="blockquote">
<h2 id="simplify-the-invocation-or-to-test-your-installation-if-you-installed-with-miniconda3" class="anchored">Simplify the Invocation, or to Test your installation if you installed with miniconda3:</h2>
<p>To simplify the invocation you could also create a script in the ~/bin folder:</p>
<pre><code>$ echo '#!/bin/bash' &gt; ~/bin/trimmomatic
$ echo 'java -jar ~/src/Trimmomatic-0.36/trimmomatic-0.36.jar $@' &gt;&gt; ~/bin/trimmomatic
$ chmod +x ~/bin/trimmomatic</code></pre>
<p>{: .bash}</p>
<p>Test your script by running:</p>
<pre><code>$ trimmomatic</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
</section>
<section id="bwa" class="level3">
<h3 class="anchored">BWA</h3>
<blockquote class="blockquote">
<h2 id="macos-3" class="anchored" data-anchor-id="bwa">MacOS</h2>
<pre><code>conda install -c bioconda bwa=0.7.17=ha92aebf_3</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="bwa-source-code-installation" class="anchored">BWA Source Code Installation</h2>
<p>If you prefer to install from source, follow the instructions below:</p>
<pre><code>$ cd ~/src
$ curl -OL http://sourceforge.net/projects/bio-bwa/files/bwa-0.7.17.tar.bz2
$ tar jxvf bwa-0.7.17.tar.bz2
$ cd bwa-0.7.17
$ make
$ export PATH=~/src/bwa-0.7.17:$PATH</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<p><strong>Test your installation by running:</strong></p>
<pre><code>$ bwa</code></pre>
<p>{: .bash}</p>
</section>
<section id="samtools" class="level3">
<h3 class="anchored">SAMtools</h3>
<blockquote class="blockquote">
<h2 id="macos-4" class="anchored" data-anchor-id="samtools">MacOS</h2>
<pre><code>$ conda install -c bioconda samtools=1.9=h8ee4bcc_1</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="samtools-versions" class="anchored">SAMtools Versions</h2>
<p>SAMtools has changed the command line invocation (for the better). But this means that most of the tutorials on the web indicate an older and obsolete usage. {: .callout}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="samtools-source-code-installation" class="anchored">SAMtools Source Code Installation</h2>
<p>If you prefer to install from source, follow the instructions below:</p>
<pre><code>$ cd ~/src
$ curl -OkL https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
$ tar jxvf samtools-1.9.tar.bz2
$ cd samtools-1.9
$ make</code></pre>
<p>{: .bash}</p>
<p>Add directory to the path if necessary:</p>
<pre><code>$ echo export `PATH=~/src/samtools-1.9:$PATH` &gt;&gt; ~/.bashrc
$ source ~/.bashrc</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<p><strong>Test your installation by running:</strong></p>
<pre><code>$ samtools</code></pre>
<p>{: .bash}</p>
</section>
<section id="bcftools" class="level3">
<h3 class="anchored">BCFtools</h3>
<blockquote class="blockquote">
<h2 id="macos-5" class="anchored" data-anchor-id="bcftools">MacOS</h2>
<pre><code>$ conda install -c bioconda bcftools=1.8=h4da6232_3 </code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="bcf-tools-source-code-installation" class="anchored">BCF tools Source Code Installation</h2>
<p>If you prefer to install from source, follow the instructions below:</p>
<pre><code>$ cd ~/src
$ curl -OkL https://github.com/samtools/bcftools/releases/download/1.8/bcftools-1.8.tar.bz2
$ tar jxvf bcftools-1.8.tar.bz2
$ cd bcftools-1.8
$ make</code></pre>
<p>{: .bash}</p>
<p>Add directory to the path if necessary:</p>
<pre><code>$ echo export `PATH=~/src/bcftools-1.8:$PATH` &gt;&gt; ~/.bashrc
$ source ~/.bashrc</code></pre>
<p>{: .bash} {: .solution}</p>
</blockquote>
<p><strong>Test your installation by running:</strong></p>
<pre><code>$ bcftools</code></pre>
<p>{: .bash}</p>
</section>
<section id="igv" class="level3">
<h3 class="anchored" data-anchor-id="igv">IGV</h3>
<ul>
<li><a href="https://software.broadinstitute.org/software/igv/download">Download the IGV installation files</a></li>
<li>Install and run IGV using the <a href="https://software.broadinstitute.org/software/igv/download">instructions for your operating system</a>.</li>
</ul>
</section>
<section id="running-fastqc" class="level2">
<h2 class="anchored" data-anchor-id="running-fastqc">Running FastQC</h2>
<p>We will now assess the quality of the reads that we downloaded. First, make sure you’re still in the <code>untrimmed_fastq</code> directory</p>
<pre><code>$ cd ~/Desktop/CLI_Class_Directory/shell_data/untrimmed_fastq/ </code></pre>
<p>{: .bash}</p>
<blockquote class="blockquote">
<h2 id="exercise" class="anchored">Exercise</h2>
<p>How big are the files? (Hint: Look at the options for the <code>ls</code> command to see how to show file sizes.)</p>
<blockquote class="blockquote">
<h2 id="solution" class="anchored">Solution</h2>
<pre><code>$ ls -l -h</code></pre>
<p>{: .bash}</p>
<pre><code>-rw-r--r--  1 jpcourneya  staff    46K Jul 27 16:43 SRR097977.fastq
-rw-r--r--  1 jpcourneya  staff    42K Jul 27 16:43 SRR098026.fastq</code></pre>
<p>{: .output}</p>
<p>There are 2 FASTQ files ranging from 42k to 46k.</p>
<p>{: .solution} {: .challenge}</p>
</blockquote>
</blockquote>
<p>FastQC can accept multiple file names as input, and on both zipped and unzipped files, so we can use the *.fastq* wildcard to run FastQC on all of the FASTQ files in this directory.</p>
<pre><code>$ fastqc *.fastq* </code></pre>
<p>{: .bash}</p>
<p>You will see an automatically updating output message telling you the progress of the analysis. When the analysis completes, your prompt will return. So your screen will look something like this:</p>
<pre><code>Started analysis of SRR097977.fastq
Analysis complete for SRR097977.fastq
Started analysis of SRR098026.fastq
Analysis complete for SRR098026.fastq</code></pre>
<p>{: .output}</p>
<p>The FastQC program has created several new files within our <code>data/untrimmed_fastq/</code> directory.</p>
<pre><code>$ ls </code></pre>
<p>{: .bash}</p>
<pre><code>SRR097977.fastq     SRR097977_fastqc.html   SRR097977_fastqc.zip    SRR098026.fastq     SRR098026_fastqc.html   SRR098026_fastqc.zip</code></pre>
<p>{: .output}</p>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code> file and a <code>.html</code> file. The <code>.zip</code> file extension indicates that this is actually a compressed set of multiple output files. We’ll be working with these output files soon. The <code>.html</code> file is a stable webpage displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we will move these output files into a new directory within our <code>results/</code> directory.</p>
<pre><code>$ mkdir -p ~/Desktop/CLI_Class_Directory/results/fastqc_untrimmed_reads 
$ mv *.zip ~/Desktop/CLI_Class_Directory/results/fastqc_untrimmed_reads/ 
$ mv *.html ~/Desktop/CLI_Class_Directory/results/fastqc_untrimmed_reads/ </code></pre>
<p>{: .bash}</p>
<p>Now we can navigate into this results directory and do some closer inspection of our output files.</p>
<pre><code>$ cd ~/Desktop/CLI_Class_Directory/results/fastqc_untrimmed_reads/ </code></pre>
<p>{: .bash}</p>
</section>
<section id="viewing-the-fastqc-results" class="level2">
<h2 class="anchored" data-anchor-id="viewing-the-fastqc-results">Viewing the FastQC results</h2>
<p>Since we were working on our local computers, we’re able to look at each of these HTML files by opening them in a web browser.</p>
</section>
<section id="assessing-quality-using-fastqc" class="level2">
<h2 class="anchored" data-anchor-id="assessing-quality-using-fastqc">Assessing Quality using FastQC</h2>
<p>In real life, you won’t be assessing the quality of your reads by visually inspecting your FASTQ files.</p>
<p>FastQC has a number of features which can give you a quick impression of any problems your data may have, so you can take these issues into consideration before moving forward with your analyses. Rather than looking at quality scores for each individual read, FastQC looks at quality collectively across all reads within a sample. The image below shows one FastQC-generated plot that indicates a very high quality sample:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://i.imgur.com/I1YS90b.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">good_quality</figcaption><p></p>
</figure>
</div>
<p>The x-axis displays the base position in the read, and the y-axis shows quality scores. In this example, the sample contains reads that are 40 bp long. This is much shorter than the reads we are working with in our workflow. For each position, there is a box-and-whisker plot showing the distribution of quality scores for all reads at that position. The horizontal red line indicates the median quality score and the yellow box shows the 1st to 3rd quartile range. This means that 50% of reads have a quality score that falls within the range of the yellow box at that position. The whiskers show the absolute range, which covers the lowest (0th quartile) to highest (4th quartile) values.</p>
<p>For each position in this sample, the quality values do not drop much lower than 32. This is a high quality score. The plot background is also color-coded to identify good (green), acceptable (yellow), and bad (red) quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the spectrum.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://i.imgur.com/GT1s2wT.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">bad_quality</figcaption><p></p>
</figure>
</div>
<p>Here, we see positions within the read in which the boxes span a much wider range. Also, quality scores drop quite low into the “bad” range, particularly on the tail end of the reads. The FastQC tool produces several other diagnostic plots to assess sample quality, in addition to the one plotted above.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>